### DISCUSSION TITLE
Concepts for numeric_types

### DISCUSSION THREAD

**Discussion by osayamenja** [NVIDIA/cutlass#1631] [Q&A] [answered]
Hello,

In a nutshell, is there a more concise way to define a C++ concept for all cutlass numeric types?

I am currently writing some C++ code using `cute` numeric types and I would like to define a concise concept that captures all its numeric types. 

For example, for some `TensorType t` using `std::same_as<t, cute::half_t> || std::same_as<t, cute::float_e4m3_t> || ... || std::same_as<t, cute::complex<half_t>> || ...` should be fine.

However, is there a fundamental type, that would perhaps allow for one shot check with e.g. `std::common_with` or `std::derived_from`?

I ask this question because in my code, I am invoking something along the lines of `t_0 = TensorType(0.0)`, which I know all `cute` numeric types support. I am also very open to suggestions that may not involve using concepts at all.

Thanks.

**Comment by osayamenja** [comment]
Forgot to mention that `is_integral` and `is_floating_point` fails a static assertion for `integer_subbyte` and cute float structs, respectively; so, this would not apply, except I am doing something completely wrong here :)

Also, I am using concepts via [libcu++ ](https://nvidia.github.io/cccl/libcudacxx/standard_api/concepts_library.html#restrictions)

**Reply by thakkarV** [reply, marked as answer]
Unfortunately, we do not have plans to move to C++20, so we can't really start using concepts anytime soon. That said, I would expect `is_integral` to just work for integer subbyte too, and if they do not its likely an oversight that we would appreciate an MR for :)

**Reply by osayamenja** [reply]
Understood. OTOH, libcu++ [backports](https://nvidia.github.io/cccl/libcudacxx/standard_api.html?highlight=backport#standard-library-backports) concepts to C++ 14; but with limited functionality as the `concepts` or `requires` keywords are unavailable.

I decided on using the simple concept below to capture my use case as it applies both to cute numeric and, more generally, to language native types.
```cpp
template<typename V>
    concept TensorValueType = requires(V v){
        {V(0)} -> std::same_as<V>;
    };
```

Regarding the MR, going off this [reference](https://en.cppreference.com/w/cpp/types/is_integral), I think making this work would require redefining the alias types for integer_subbyte, that is [these](https://github.com/NVIDIA/cutlass/blob/56b46e2d13875b46b8f6a03f9f5ac91e2bfdc01a/include/cutlass/integer_subbyte.h#L154C1-L167C44) would become, for example, `using uint1b_t = decltype(integer_subbyte<1, false>.Storage)`

What do you think? @thakkarV

**Reply by thakkarV** [reply]
I do not think it would require new aliases -- instead you can just define new specializations for `is_integral` for the subbyte ints

```cpp
struct cute::is_integral<cutlass::uint1b_t> : true_type {};
```

**Comment by osayamenja** [comment]
Side-note, how do we get the underlying value type of a tensor? I use below but is there a simpler way?
```cpp
// For cute::Tensor input
using TensorValueType = typename cute::iterator_traits<decltype(input.data())>::value_type;
```

**Reply by thakkarV** [reply]
https://github.com/NVIDIA/cutlass/blob/main/include/cute/tensor.hpp#L110

should be what you want

**Reply by osayamenja** [reply]
For the above snippet, would this suggest replacing `input.data()` with `input.engine().begin()`? If not, could you kindly give an example?

**Reply by thakkarV** [reply]
simpler still

```cpp
using ValueType = typename decltype(tensor)::value_type;
```

**Reply by osayamenja** [reply]
thanks!! @thakkarV
